rfcs/0074-tvmscript-unified-printer.md

```diff
@@ -0,0 +1,475 @@
+propose to develop the infrastructure to enable distributed registration, and
+further allows printer for different levels of IR to be registered in separate
+translation units, and in the meantime keeps the capability to be mixed
+together at various level, for example, Relax uses TIR expression in its
+function bodies, and TIR calls back to Relax function.
+
+## Diagnostics and Reproducibility
+
+Existing error reporting mechanisms have not taken IR structure and
+reproducibility into consideration. Usually it reports a single line error
+message without providing necessary context of how the IR looks like during
+compilation. For example, when comparing whether two TIRs are structurally
+equivalent, the system may report:
+
+```cpp
+ValueError: StructuralEqual check failed, caused by lhs:
+{slow_memory_3_var: buffer(slow_memory_3_buffer_var, 0x501bf80), fast_memory_2_var: buffer(fast_memory_2_buffer_var, 0x501bd80), placeholder_3: buffer(placeholder_5, 0x50138a0), placeholder_2: buffer(placeholder_4, 0x5012b60), T_subtract: buffer(T_subtract_1, 0x5014390)}
+and rhs:
+{}
+```
+
+which lacks necessary information for users to understand where the mismatch
+is.
+
+As a recent effort, structural error reporting in TIR scheduling provides
+relevant and reproducible context, as demonstrated below:
+
+```cpp
+@tvm.script.ir_module
+class Module:
+    @tir.prim_func
+    def main(a: tir.handle, b: tir.handle) -> None:
+        A = tir.match_buffer(a, [128, 128, 128, 128], dtype="float32")
+        B = tir.match_buffer(b, [128, 128, 128, 128], dtype="float32")
+        # body
+        # with tir.block("root")
+        for i, j, k, l in tir.grid(128, 128, 128, 8):
+            tir.Block#0
+            with tir.block("B"):
+            ^^^^^^^^^^^^^^^^^^^^
+                vi, vj, vk = tir.axis.remap("SSS", [i, j, k])
+                vl = tir.axis.spatial(128, l * 16)
+                tir.reads([A[vi, vj, vk, vl]])
+                tir.writes([B[vi, vj, vk, vl]])
+                B[vi, vj, vk, vl] = A[vi, vj, vk, vl] * tir.float32(2)
+
+Error: ...
+```
+
+However, the underlying mechanism supports only S-TIR and error reporting on
+`tir.ForNode` and `tir.BlockNode`, and is less extensible for generic cases.
+
+To generalize this UX across the TVM stack, during the first stage in
+translation, the following steps is additionally executed:
+
+- Each Doc node is optionally attached to a node in TVM IR
+- After the 1st stage is finished, collect all IR nodes that gets attached to
+  Doc into a map, whose key is IR node and value is a list of Doc nodes.
+- For each IR node that has diagnostic message, trace back through its parent
+  until it reaches to an IR node in the map collected in previous step. Then it
+  can produce a map from Doc node to diagnostic message.
+- In the 2nd stage, diagnostic message will be printed as doc is being printed
+  into target language
+
+# Reference-level explanation
+
+## Doc Spec
+
```

Comment:
How to represent the type and nested types in the doc nodes ?

Reply:
Types will be represented by `ExprDoc`. For example, `Tensor((2, 3), "float32")` will be 

```
CallDoc(
  IdDoc("Tensor"),
  [
    TupleDoc([
       LiteralDoc(2),
       LiteralDoc(3)
    ]),
    LiteralDoc("float32")
  ]
)
```